README.md

```diff
@@ -237,6 +237,180 @@ You can do the same for Kafka Streams and Kafka, by using the `kstreams` and `ka
+
+This section is not meant to be an exhaustive decription of the harness file.  New tutorial authors should not need to create a harness file from scratch, using either the `tools/gen_project.sh` or  `tools/clone.sh` script will provide a usable harness file.  This section should provide enough guidance to add, update, or remove sections as needed.
+
+#### TLDR;
+
+To run a tutorial programatically do the following stepes. Note these instructions assume you've already checked out the kafka-tutorials repo).
+
+1. cd into `_includes/tutorials/<tutorial name>/<type>/code` directory where type is one of `ksql | kstreams | kafka`.
+2. run the `make` command.
+
+The `Makefile` will delete and re-create the `outputs` directory used contain files with output from various steps used to verify the tutorial steps.
+
+Here's an example of the contents of a `Makefile`:
+
+```yml
+STEPS_DIR := tutorial-steps
+DEV_OUTPUTS_DIR := $(STEPS_DIR)/dev/outputs
+TEMP_DIR := $(shell mktemp -d)
+
+tutorial:
+  rm -r $(DEV_OUTPUTS_DIR) || true
+  mkdir $(DEV_OUTPUTS_DIR)
+  harness-runner ../../../../../_data/harnesses/fk-joins/kstreams.yml $(TEMP_DIR)
+  diff --strip-trailing-cr $(STEPS_DIR)/dev/expected-output-events.json $(DEV_OUTPUTS_DIR)/music-interest.json
+```
+
+The last line uses the `diff` command to validate expected output against the actual output of tutorial steps.  The `Makefile` may have more than one valdiation step so it will have separate `diff` commmands for each verification.  For example:
+
+```yml
+STEPS_DIR := tutorial-steps
+DEV_OUTPUTS_DIR := $(STEPS_DIR)/dev/outputs
+TEMP_DIR := $(shell mktemp -d)
+
+tutorial:
+  rm -r $(DEV_OUTPUTS_DIR) || true
+  mkdir $(DEV_OUTPUTS_DIR)
+  harness-runner ../../../../../_data/harnesses/dynamic-output-topic/kstreams.yml $(TEMP_DIR)
+  diff --strip-trailing-cr $(STEPS_DIR)/dev/expected-output.json $(DEV_OUTPUTS_DIR)/actual-output.json
+  diff --strip-trailing-cr $(STEPS_DIR)/dev/expected-special-output.json $(DEV_OUTPUTS_DIR)/actual-special-order-output.json
+```
+
+
+
+
+#### 1. Structure
+
+Three top-level sections make up the harness file:
+
+* dev - the setup and teaching portion of the tutorial (required)
+* test - test setup and execution of tests, if any (optional)
+* prod - steps to build and deploy a docker image of the tutorial code (optional)
+
+In some cases, having a test and or prod section doesn't make sense, so those sections can be left out of the harness file.  The Apache Kafka console producer and consumer tutorials are a good example of tutorials that don't need a test or prod section.
```

Comment:
Can you add a hyperlink to those tutorials?

Reply:
ack